### PR TITLE
XLATensor convolution and shape inference infrastructure 

### PR DESCRIPTION
**Pull request by asuhan**

### CODE REVIEW

File: torch_xla/csrc/ops/conv2d.cpp [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+      precision_(MakePrecisionConfig(use_full_conv_precision)) {}
+
+Conv2d::Conv2d(const NodeOperand& input, const NodeOperand& weight, int stride,
+               int padding, bool use_full_conv_precision)
+    : Node(ir::OpKind(at::aten::convolution), {input, weight},
+           NodeOutputShape(input, weight, stride, padding)),
+      stride_(stride),
+      padding_(padding),
+      precision_(MakePrecisionConfig(use_full_conv_precision)) {}
+
```
Comment (dlibenzi) [inline torch_xla/csrc/ops/conv2d.cpp]: Add a ToString().

File: torch_xla/csrc/ops/conv2d.h [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+// IR node for 2D convolutions with or without bias.
+class Conv2d : public Node {
+ public:
+  Conv2d(const NodeOperand& input, const NodeOperand& weight,
+         const NodeOperand& bias, int stride, int padding,
+         bool use_full_conv_precision);
+
+  Conv2d(const NodeOperand& input, const NodeOperand& weight, int stride,
+         int padding, bool use_full_conv_precision);
+
```
Comment (dlibenzi) [inline torch_xla/csrc/ops/conv2d.h]: ToString().

File: torch_xla/csrc/ops/conv2d.h [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+  XlaOpVector Lower(LoweringContext* loctx) const override;
+
+ private:
+  // The parameters of the convolution. Only support the same stride and padding
+  // in both dimension for now.
+  const int stride_;
```
Comment (dlibenzi) [inline torch_xla/csrc/ops/conv2d.h]: I am not totally opposed to const here, but we have a lot of code that does not have them, although constants.
This reminds me the Java's final a bit.
I'd keen to not have those.

File: torch_xla/csrc/ops/infer_output_shape.cpp [hunk 4]
```diff
@@ -0,0 +1,25 @@
+#include "ops/infer_output_shape.h"
+#include "helpers.h"
+
+namespace torch_xla {
+namespace ir {
+namespace ops {
+
+xla::Shape InferOutputShape(
+    tensorflow::gtl::ArraySlice<const xla::Shape> input_shapes,
+    LowerForShapeFn core_lowering_fn) {
```
Comment (dlibenzi) [inline torch_xla/csrc/ops/infer_output_shape.cpp]: const LowerForShapeFn&

File: torch_xla/csrc/ops/infer_output_shape.cpp [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+  xla::XlaBuilder b("infer_output_shape");
+  std::vector<xla::XlaOp> parameters;
+  for (size_t parameter_number = 0; parameter_number < input_shapes.size();
+       ++parameter_number) {
+    parameters.push_back(
+        xla::Parameter(&b, parameter_number, input_shapes[parameter_number],
+                       "param_" + std::to_string(parameter_number)));
```
Comment (dlibenzi) [inline torch_xla/csrc/ops/infer_output_shape.cpp]: I'd use StrCat here.

File: torch_xla/csrc/ops/infer_output_shape.h [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+namespace torch_xla {
+namespace ir {
+namespace ops {
+
+using LowerForShapeFn = std::function<xla::XlaOp(
+    tensorflow::gtl::ArraySlice<const xla::XlaOp> operands)>;
+
+// Compute the output shape for the given input shapes and lowering.
+xla::Shape InferOutputShape(
+    tensorflow::gtl::ArraySlice<const xla::Shape> input_shapes,
```
Comment (dlibenzi) [inline torch_xla/csrc/ops/infer_output_shape.h]: I'd take const Shape* here.
Shape is heavy, and is becoming heavier.
Maybe.
I don't know.
The XLA builder is not light either.

File: torch_xla/csrc/ops/conv2d.h (hunk 3 above)
Comment (asuhan) [reply torch_xla/csrc/ops/conv2d.h]: I've removed them, although I think we should instead fix the rest of the code to have them.

File: torch_xla/csrc/ops/infer_output_shape.h (hunk 6 above)
Comment (asuhan) [reply torch_xla/csrc/ops/infer_output_shape.h]: That'd make the caller a mess, let's not do it until we know it's a problem. Also, the slice shouldn't copy the elements, so it's not *that* bad.

Comment (dlibenzi) [reply torch_xla/csrc/ops/infer_output_shape.h]: It does copy. The std::initializer_list<> does.
Wouldn't be a big mess. Just prepend an & in front.
If you want to leave it as is I am OK.
But it's going to be a PITA to change all the call site if it turns out to be.

**Pull request merged**